sdk/cosmosdb/cosmos/src/queryExecutionContext/EndpointComponent/OffsetLimitEndpointComponent.ts

```diff
@@ -22,18 +25,55 @@ export class OffsetLimitEndpointComponent implements ExecutionContext {
     const buffer: any[] = [];
     const response = await this.executionContext.fetchMore(diagnosticNode);
     mergeHeaders(aggregateHeaders, response.headers);
-    if (response === undefined || response.result === undefined) {
-      return { result: undefined, headers: response.headers };
+    if (
+      response === undefined ||
+      response.result === undefined ||
+      !Array.isArray(response.result.buffer) ||
+      response.result.buffer.length === 0
+    ) {
+      const result = createParallelQueryResult([], new Map(), {}, undefined);
+
+      return { result, headers: response.headers };
     }
 
-    for (const item of response.result) {
+    const parallelResult = response.result as ParallelQueryResult;
+    const dataToProcess: any[] = parallelResult.buffer;
+    const partitionKeyRangeMap = parallelResult.partitionKeyRangeMap;
+    const updatedContinuationRanges = parallelResult.updatedContinuationRanges;
+    const orderByItems = parallelResult.orderByItems;
+
+    const initialOffset = this.offset;
+    const initialLimit = this.limit;
+
+    for (const item of dataToProcess) {
       if (this.offset > 0) {
         this.offset--;
       } else if (this.limit > 0) {
         buffer.push(item);
         this.limit--;
       }
     }
-    return { result: buffer, headers: aggregateHeaders };
+
+    // Process offset/limit logic and update partition key range map
+    // Note: Pass initial offset/limit values (not current state) to calculateOffsetLimitForPartitionRanges
+    // This function updates partition metadata while the loop above processes actual data items
```

Comment:
Isn't it better to do bpoth at same place? 
1. Move calculateOffsetLimitForPartitionRanges to this class.
2. While updating partition ranges, can we process data to remain consistent?